### PR TITLE
🐛 delete port left over by err openstackmachine

### LINKED ISSUE #1254: delete port left over by error openstackmachine
/kind bug

**What steps did you take and what happened:**
[A clear and concise description of what the bug is.]
This is an abnormal scenario: the VM creation fails, but the port creation succeeds. In this case, the port will be left when the cluster  is deleted

**What did you expect to happen:**
In this case,we could delete port by subnet

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Cluster API Provider OpenStack version (Or `git rev-parse HEAD` if manually built):
- Cluster-API version:
- OpenStack version:
- Minikube/KIND version:
- Kubernetes version (use `kubectl version`):
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by ruifaling**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1254 

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**TODOs**:

- [ ] squashed commits
- if necessary:
  - [ ] includes documentation
  - [ ] adds unit tests

/hold

### CODE REVIEW

**Comment by tobiasgiese** [comment]
/ok-to-test

Thanks for contributing @ruifaling 🙂 
Could you please extend your PR description or create a proper issue regarding your bug? Thanks!

**Comment by tobiasgiese** [comment]
/ok-to-test

Thanks for contributing @ruifaling 🙂 
Could you please extend your PR description or create a proper issue regarding your bug? Thanks!

**Comment by tobiasgiese** [comment]
/retitle 🐛 delete port left over by err openstackmachine

**Comment by tobiasgiese** [comment]
@ruifaling according to the `pull-cluster-api-provider-openstack-test` you should check the linter errors

```
pkg/cloud/services/networking/network.go:20: File is not `gofumpt`-ed (gofumpt)
	"fmt"
pkg/cloud/services/networking/network.go:22: File is not `gofumpt`-ed (gofumpt)
	"strings"
pkg/cloud/services/networking/network.go:21: File is not `goimports`-ed (goimports)
	capoerrors "sigs.k8s.io/cluster-api-provider-openstack/pkg/utils/errors"
pkg/cloud/services/networking/network.go:154:2: var-naming: var networkId should be networkID (revive)
	networkId := openStackCluster.Spec.Network.ID
	^
pkg/cloud/services/networking/network.go:172:60: ST1017: don't use Yoda conditions (stylecheck)
		if strings.Contains(port.Name, openStackCluster.Name) && "" == port.DeviceOwner {
		                                                         ^
```

**Comment by ruifaling** [comment]
> @ruifaling according to the `pull-cluster-api-provider-openstack-test` you should check the linter errors
> 
> ```
> pkg/cloud/services/networking/network.go:20: File is not `gofumpt`-ed (gofumpt)
> 	"fmt"
> pkg/cloud/services/networking/network.go:22: File is not `gofumpt`-ed (gofumpt)
> 	"strings"
> pkg/cloud/services/networking/network.go:21: File is not `goimports`-ed (goimports)
> 	capoerrors "sigs.k8s.io/cluster-api-provider-openstack/pkg/utils/errors"
> pkg/cloud/services/networking/network.go:154:2: var-naming: var networkId should be networkID (revive)
> 	networkId := openStackCluster.Spec.Network.ID
> 	^
> pkg/cloud/services/networking/network.go:172:60: ST1017: don't use Yoda conditions (stylecheck)
> 		if strings.Contains(port.Name, openStackCluster.Name) && "" == port.DeviceOwner {
> 		                                                         ^
> ```

File: controllers/openstackcluster_controller.go [hunk 1]
```diff
@@ -142,6 +142,11 @@ func reconcileDelete(ctx context.Context, scope *scope.Scope, patchHelper *patch
 
 	clusterName := fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)
 
+	if err = networkingService.DeletePorts(openStackCluster); err != nil {
+		handleUpdateOSCError(openStackCluster, errors.Errorf("failed to delete ports: %v", err))
+		return reconcile.Result{}, errors.Errorf("failed to delete ports: %v", err)
```
Comment (seanschneeweiss) [inline controllers/openstackcluster_controller.go:147]:
```suggestion
		return reconcile.Result{}, errors.Wrap(err, "failed to delete ports")
```

File: pkg/cloud/services/networking/network.go [hunk 2]
```diff
@@ -147,6 +150,33 @@ func (s *Service) ReconcileNetwork(openStackCluster *infrav1.OpenStackCluster, c
...
+	networkID := openStackCluster.Spec.Network.ID
+	portList, err := s.client.ListPort(ports.ListOpts{
+		NetworkID:   networkID,
+		DeviceOwner: "",
+	})
+	if err != nil {
+		if capoerrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("searching for no owner port error: %v", err)
```
Comment (seanschneeweiss) [inline pkg/cloud/services/networking/network.go:163]: The error message is more clear if the actual request type is part of the message, e.g. "list".
```suggestion
		return fmt.Errorf("list ports of network %q: %v", networkID, err)
```

File: pkg/cloud/services/networking/network.go [hunk 3]
```diff
@@ -147,6 +150,33 @@ func (s *Service) ReconcileNetwork(openStackCluster *infrav1.OpenStackCluster, c
...
+	}
+
+	for _, port := range portList {
+		if strings.Contains(port.Name, openStackCluster.Name) && port.DeviceOwner == "" {
```
Comment (seanschneeweiss) [inline pkg/cloud/services/networking/network.go:167]: As the portList should only include ports without a device owner, omitting the condition here should be okay.
```suggestion
		if strings.Contains(port.Name, openStackCluster.Name) {
```

File: pkg/cloud/services/networking/network.go [hunk 4]
```diff
@@ -147,6 +150,33 @@ func (s *Service) ReconcileNetwork(openStackCluster *infrav1.OpenStackCluster, c
...
+			err := s.DeletePort(openStackCluster, port.ID)
+			if capoerrors.IsNotFound(err) {
+				continue
+			} else {
+				return fmt.Errorf("delete port %s of network %q failed : %v", port.ID, networkID, err)
+			}
```
Comment (seanschneeweiss) [inline pkg/cloud/services/networking/network.go:173]: If the error is nil, we would also run into the `else` case.

Maybe rewrite it to something like this:
```go
if err := s.DeletePort(openStackCluster, port.ID); err != nil {
    if capoerrors.IsNotFound(err) {
        continue
    }
    return fmt.Errorf("delete port %s of network %q failed : %v", port.ID, networkID, err)
}
```

**Review by seanschneeweiss — COMMENTED** [review]
Thank you very much @ruifaling for the fix.

Can you move the code added to `pkg/cloud/services/networking/network.go` to the file `pkg/cloud/services/networking/port.go`?

File: pkg/cloud/services/networking/network.go (hunk 4 above)
Comment (ruifaling) [reply pkg/cloud/services/networking/network.go]: Thanks for your suggestions,I have fixed them,please have a re-check

File: pkg/cloud/services/networking/port.go [hunk 5]
```diff
@@ -243,6 +244,32 @@ func (s *Service) DeletePort(eventObject runtime.Object, portID string) error {
...
+	})
+	if err != nil {
+		if capoerrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("list ports of network %q: %v", networkID, err)
+	}
+
+	for _, port := range portList {
+		if strings.Contains(port.Name, openStackCluster.Name) {
```
Comment (tobiasgiese) [inline pkg/cloud/services/networking/port.go:261]:
```suggestion
		if strings.HasPrefix(port.Name, openStackCluster.Name) {
```

`HasPrefx` will be enough as the port name ist prefixed by the instance name (and the instance name is prefixed by the cluster name :))

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/7a8c2e82b58c66ca9a20bbf22be916c956b51998/pkg/cloud/services/compute/instance.go#L279-L285

Comment (ruifaling) [reply pkg/cloud/services/networking/port.go]: yes, thanks, I have changed it,please have a re-check

File: controllers/openstackcluster_controller.go [hunk 6]
```diff
@@ -142,6 +142,11 @@ func reconcileDelete(ctx context.Context, scope *scope.Scope, patchHelper *patch
 
 	clusterName := fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)
 
+	if err = networkingService.DeletePorts(openStackCluster); err != nil {
+		handleUpdateOSCError(openStackCluster, errors.Errorf("failed to delete ports: %v", err))
```
[the remark this answers is not in the document]
Comment (tobiasgiese) [reply controllers/openstackcluster_controller.go:146]: Okay. We don't need the stack trace in the status, that's correct

**Comment by ruifaling** [comment]
@tobiasgiese @seanschneeweiss  Are there any other errors I need to correct or Is there anything others need I to do?

**Comment by ruifaling** [comment]
need I remove the lable **hold**？

**Comment by tobiasgiese** [comment]
/approve
/hold cancel

sorry for the delay @ruifaling, missed your PR 🙂

**Comment by ruifaling** [comment]
is thers something wrong with the e2e test?

**Pull request merged**